Improve ServerSideRender Component to retain preview of the component while it is loading new state

fabiankaegy opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
The `ServerSideRender` component now continues to show the markup from the previous request when it is fetching the new markup. The reason behind it is preventing the flashing of content when attributes are being changed. This work is a continuation building on the work of @dgwyer here: https://github.com/dgwyer/server-side-render-x

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
This was tested using the bundled `wp-env` environment with the `Archives` block. 

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

- ServerSideRender Component continues to show markup while loading new markup when for example attributes change


## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

fabiankaegy commented:
@dgwyer I'd love to get your thoughts on this first because I used your https://github.com/dgwyer/server-side-render-x repository as an inspiration for the change.

ZebulanStanphill commented on packages/server-side-render/src/server-side-render.js:
```diff
@@ -20,126 +20,105 @@ export function rendererPath( block, attributes = null, urlQueryArgs = {} ) {
 		const urlAttributes = isPostRequest ? null : attributes;
 		const path = rendererPath( block, urlAttributes, urlQueryArgs );
 		const data = isPostRequest ? { attributes } : null;
+		let currentFetchRequest = null;
```
This needs to be a ref at the component level, not a variable local to the current instance of the `fetchData` callback.

fabiankaegy replied:
addressed  👍

ZebulanStanphill commented on packages/server-side-render/src/server-side-render.js:
```diff
@@ -20,126 +20,105 @@ export function rendererPath( block, attributes = null, urlQueryArgs = {} ) {
-					this.isStillMounted &&
-					fetchRequest === this.currentFetchRequest &&
-					response
+					isMounted &&
```
This (and several other places) should be checking `isMounted.current` in order to actually read the stored value.

fabiankaegy replied:
solved  👍

ZebulanStanphill commented on packages/server-side-render/src/server-side-render.js:
```diff
@@ -20,126 +20,105 @@ export function rendererPath( block, attributes = null, urlQueryArgs = {} ) {
-				{ response }
-			</RawHTML>
+			<LoadingResponsePlaceholder>
+				<RawHTML key="html" className={ className }>
```
I'm pretty sure the `key` prop here serves no purpose, so this is a good opportunity to remove it.

fabiankaegy replied:
Agreed and removed  👍

ZebulanStanphill commented on packages/server-side-render/src/server-side-render.js:
```diff
@@ -20,126 +20,105 @@ export function rendererPath( block, attributes = null, urlQueryArgs = {} ) {
+	);
 }
 
 ServerSideRender.defaultProps = {
```
[`defaultProps` may disappear from React in the future](https://github.com/reactjs/rfcs/pull/107), and it's not necessary for function components anyway since those can just use default arguments. So I'd recommend not using it.

fabiankaegy replied:
@ZebulanStanphill agreed and done  👍

ZebulanStanphill commented on packages/server-side-render/src/server-side-render.js:
```diff
@@ -20,126 +20,105 @@ export function rendererPath( block, attributes = null, urlQueryArgs = {} ) {
 
-	fetch( props ) {
-		if ( ! this.isStillMounted ) {
+	useEffect( () => {
+		fetchData( props );
+	}, [ props ] );
```
This effect will run on every render because objects in the dependency array are not deep-equality-checked, and so since they are recreated on every render, `props` will always be different because it's technically a new object every time. The only way I know of to avoid re-fetching on every render is to use the `compose` package's `usePrevious` hook and Lodash's `isEqual` function like so:

```suggestion
	const prevProps = usePrevious( props );
	useEffect( () => {
		if ( ! isEqual( prevProps, props ) ) {
			fetchData( props );
		}
	} );
```

ZebulanStanphill commented on packages/server-side-render/src/server-side-render.js:
```diff
@@ -20,126 +20,105 @@ export function rendererPath( block, attributes = null, urlQueryArgs = {} ) {
+		fetchData( props );
+	}, [ props ] );
+
+	function fetchData( options ) {
```
Passing in options here isn't really necessary, because this function is already within the scope of the parent component, which means we can just use the props directly.

fabiankaegy replied:
Agreed and done  👍

ZebulanStanphill reviewed (changes requested):
Coincidentally, I happened to be working on a PR that just refactors `ServerSideRender` into a function component using React hooks: #28297. (Which your PR also does, in addition to updating the behavior.) I didn't realize that this PR existed until after I'd created mine. :stuck_out_tongue:

It might make more sense for my PR to be reviewed and merged first since it would make this one a lot smaller and easier to review. Either way, feel free to rebase this PR off of mine or just use it as a guide to improve your own. I've managed to spot several issues/potential-tweaks that I've already addressed in my PR.

fabiankaegy commented:
@ZebulanStanphill Thanks for the review and sorry that you were also working on this at the same time :( 

I'm fine either way. I have addressed all your comments and tested it locally. :) So should be in a better state now 👍

ZebulanStanphill reviewed (dismissed):
One of the tests is failing because `package-lock.json` needs to be updated. You can do this by running `npm install`.

fabiankaegy commented on packages/server-side-render/src/server-side-render.js:
```diff
@@ -20,39 +21,40 @@ export function rendererPath( block, attributes = null, urlQueryArgs = {} ) {
-	componentDidUpdate( prevProps ) {
-		if ( ! isEqual( prevProps, this.props ) ) {
-			this.fetch( this.props );
+		fetchData = debounce( fetchData, 500 );
```
Ahh you are correct. I have updated my code to also use `useDebounce` in the same way.

fabiankaegy commented on packages/server-side-render/src/server-side-render.js:
```diff
@@ -69,99 +71,85 @@ export class ServerSideRender extends Component {
+	);
+}
 
 export default ServerSideRender;
```
Solved  👍

fabiankaegy commented on packages/server-side-render/src/server-side-render.js:
```diff
@@ -69,99 +71,85 @@ export class ServerSideRender extends Component {
 
-ServerSideRender.defaultProps = {
-	EmptyResponsePlaceholder: ( { className } ) => (
+export function DefaultEmptyResponsePlaceholder( { className } ) {
```
Yeah you are right  👍 removed  👍

fabiankaegy commented:
@ZebulanStanphill Thank you for taking the time to review all of this. I have updated all the requested changes :)

fabiankaegy commented on packages/server-side-render/src/server-side-render.js:
```diff
@@ -20,39 +21,39 @@ export function rendererPath( block, attributes = null, urlQueryArgs = {} ) {
+	const prevProps = usePrevious( props );
+	useEffect( () => {
+		if ( ! isEqual( prevProps, props ) ) {
+			debouncedFetchData();
```
Thanks for the detailed info. I have made the update accordingly :)

Mamaduka commented:
Hi, @fabiankaegy

Do you have time to rebase the branch off of the current `trunk`?

fabiankaegy commented:
@Mamaduka Looking at it now  👍

fabiankaegy committed:
make ServerSideRender component show previous results while fetching new data

fabiankaegy committed:
fix store previous markup state to render loading state

Mamaduka commented:
Thanks for the follow-up, @fabiankaegy.

It looks like I missed the notification, but I will try to review this tomorrow.

Mamaduka commented on packages/server-side-render/src/server-side-render.js:
```diff
@@ -62,6 +71,7 @@ export default function ServerSideRender( props ) {
 	const isMountedRef = useRef( true );
 	const fetchRequestRef = useRef();
 	const [ response, setResponse ] = useState( null );
+	const previousResponse = usePrevious( response );
```
```suggestion
	const prevResponse = usePrevious( response );
```

Just a nitpick: Maybe we should match the naming pattern here.

Mamaduka commented on packages/server-side-render/src/server-side-render.js:
```diff
@@ -39,11 +39,20 @@ function DefaultErrorResponsePlaceholder( { response, className } ) {
 }
 
-function DefaultLoadingResponsePlaceholder( { className } ) {
+function DefaultLoadingResponsePlaceholder( { children } ) {
```
I think removing the `className` prop will make this a breaking change. We should avoid that.

Mamaduka reviewed (commented):
Thanks again for working on this, @fabiankaegy.

Changes look great, and I left few minor comments regarding the code.

Here is an example of the updated UX:

https://user-images.githubusercontent.com/240569/131334636-8001065f-bdf9-4fb0-b15b-dedfad58f374.mp4

One thing I'm not sure about is hardcoded inline styles. It will make it harder for developers to override them without providing a custom `LoadingResponsePlaceholder` component. But since the package doesn't ship with styles file, I guess this is our only option.

P.S. Maybe we should match loading state styling to blocks design. Video Block example:

https://user-images.githubusercontent.com/240569/131335086-3f471cf9-2cc1-45ae-a72e-989a4c51ba10.mp4

@jasmussen, any recommendations regarding the styling of the loading state?

jasmussen commented:
> any recommendations regarding the styling of the loading state?

Thanks for the ping. Overall I think loading states across the board need a holistic rethink. The spinner needs love, and the states themselves could be looked at on whether they need spinners in the first place: if the loading happens fast enough, it's best not to show a spinner at all (i.e. the spinner could appear after a delay).

However the current loading state for the RSS feed is so mediocre that what's shipping already in this branch is an improvement, and since we need to look at loading states holistically, probably best to not hold up this PR for that.

I do share your instinct, though, that if there's prior art, it'd be nice to leverage that, and the prior art for the video is a good one to emulate. I.e. the entire item that is in a loading state becomes semi-transparent, and the spinner is overlaid dead center of the container. Same as for the image:

![image](https://user-images.githubusercontent.com/1204802/131336139-0254d1e7-e281-47a7-a658-7df709ee4ae6.gif)

Mamaduka commented:
Thanks for your input, @jasmussen. It's much appreciated as always.

Mamaduka committed:
Merge branch 'trunk' into feature/improve-server-side-render

Mamaduka committed:
Add className prop back to loading placeholder

Mamaduka committed:
Standardize 'prev' variable naming

Mamaduka committed:
Leverage existing loading styles

Mamaduka commented:
I updated the loading state styling to match existing ones in the core.

@fabiankaegy, @jasmussen, I would love to get your feedback if you think this is in a good state to land.

https://user-images.githubusercontent.com/240569/133982413-4d7d1033-4981-4a8d-a7d7-43b3b661bddb.mp4

fabiankaegy commented on packages/server-side-render/src/server-side-render.js:
```diff
@@ -39,11 +39,22 @@ function DefaultErrorResponsePlaceholder( { response, className } ) {
-		<Placeholder className={ className }>
-			<Spinner />
-		</Placeholder>
+		<div className={ className } style={ { position: 'relative' } }>
```
@Mamaduka The reason why I removed the `className` reference here is because it also is included here: https://github.com/WordPress/gutenberg/pull/28289/files#diff-1fcc38b89edc0d5f7b6239481f42d74b8e2010d792fb26b1e0f8623b222a06fbR160 meaning we now have markup where we nest the className in itself. And I worried that this would lead to stying conflicts. 

Should we maybe do something like `${className}__loading-placeholder` ?

Mamaduka replied:
I a good call, and sorry I missed the obvious clue 😄 

The `DefaultLoadingResponsePlaceholder` also received a complete set of props, so if someone is overriding the default placeholder, this change won't break anything.

I'm going to revert the change.

fabiankaegy commented:
@Mamaduka I think that visual style works and is a great improvement over the current styling because we don't have that huge visual jump. I would be okay with shipping it like that. :) Thank you for your help! :)

Mamaduka committed:
Remove className

jasmussen commented:
From a quick glance at the video (thank you for providing!) I see the spinner centered in the box, which in its transient state is semi-transparent. Which seems good to me, thank you!

dgwyer commented:
Good to see some progress on this. However, I'm not really a fan of the 'dimming' of the block content during the re-render cycle. I can appreciate _why_ it was done this way. i.e. To make it easier to place the spinner in a consistent position central to the block.

Dimming though, introduces unecessary visual cruft for me. Especially for multiple quick fire changes in the block settings panel. I originally created [my own version](https://wpgoplugins.com/an-improved-server-side-render-component-for-dynamic-wordpress-blocks/) of the SSR component to _remove_ as much visual change between the new rendered content and the previous content as possible.

e.g. This is my initial original effort. See how even for rapid settings changes the UI changes look almost as smooth as if it was being updated directly via JS.

![](https://wpgoplugins.com/wp-content/uploads/2020/08/tweak-atts-new.gif)

Mamaduka commented:
Thanks for the feedback, @dgwyer.

I tried without dimming, but then Spinner isn't visible in some cases.

After this change, you should be able to override the `LoadingResponsePlaceholder` prop with a custom component and use styling that matches your block's UI. In addition, there will be no need to maintain the custom `ServerSideDrender` component since now you have access to the `prevResponse` as `children` prop.

jasmussen commented:
Perhaps a separate exploration can be to only show the dimming _and_ the spinner after a delay of a second or so, in an optimistic assumption that the operation will have finished before that.

dgwyer commented:
@Mamaduka As I mentioned, it's obvious why the dimming was introduced but it doesn't really help solve the original issue. Ideally, no visual changes to the content transition should be introduced apart from the spinner. This was the primary motivation behind the original implementation.

I [already addressed](https://wpgoplugins.com/an-improved-server-side-render-component-for-dynamic-wordpress-blocks/) the issue of the spinner not being visible in some cases by providing a `spinnerLocation` prop to adjust the spinner relative location as required. Perhaps we can implement something similar here rather than introduce unecessary additional UI?

dgwyer commented:
@jasmussen Debouncing might work. But I think being able to adjust the spinner location should be fine in almost all cases?

Mamaduka commented:
@dgwyer I would like to avoid controlling individual pieces of the component using props. It could easily increase the complexity of the component.

It's a little more code, but you could achieve the same results by overriding default loading placeholder:
```js
function LoadingResponsePlaceholder( { children } ) {
	return (
		<div style={ { position: 'relative' } }>
			<div
				style={ {
					position: 'absolute',
					top: '10px',
					righ: '0',
				} }
			>
				<Spinner />
			</div>
			{ children }
		</div>
	);
}

<ServerSideRender
	block="core/archives"
	attributes={ attributes }
	LoadingResponsePlaceholder={ LoadingResponsePlaceholder }
/>
```

fabiankaegy commented:
@Mamaduka Is there anything left that we need to do before this can get merged? Currently Merging is still blocked beause it needs approval.

dgwyer commented:
@fabiankaegy See my comments above.

fabiankaegy commented:
@dgwyer I saw your note but thought that it was addressed by the feedback from @Mamaduka here: https://github.com/WordPress/gutenberg/pull/28289#issuecomment-922842458 

Because the `ServerSideRender` component now takes a LoadingPlaceholder component that gives you the ability to customize the experience to your liking this should give the flexibility to achieve the experience you are going after in core.

jasmussen commented:
> Debouncing might work. But I think being able to adjust the spinner location should be fine in almost all cases?

The postponed spinner feels like it could potentially benefit a lot of these blocks in a generic and reproducible way. The problem with spinner location is that you can't be sure any location is "safe", as it depends on the content inside and how it wraps. Since this PR is an improvement over what's shipping, it feels worth landing it and continuing conversation in new PRs and tickets.

dgwyer commented:
The point is the **defaut behaviour** shouldn't have any additional layering over existing content during content re-rendering. The 'problem' of spinner location is easily mitigated by displaying it outside the block content. I've been doing this since the beginning and it just works (very well).

Adding UI on top of the content during render transition misses the point entirely. For JS only blocks the content is seamlessly updated, and that's what we want to achieve with SSR blocks too. My solution displays the spinner so it does not intefere with block content at all. See the GIF above (I've since update it so it displays to the left of the block container, vertically centered). This is a much better rendering experience for the user.

I've been using this daily for well over a year and have seen thousands of transitions. Dimming the content and adding the spinner right in the center of the block content during a render transition is not an ideal UX.

dgwyer commented:
@jasmussen Outide of the block content can be considered a safe location.

I've been using this for a long time and it works very well. As you can see, rapid updating of attributes renders block updates very smoothly. Having the content blinking on/off continuously makes this a very poor UX and is why I implemented this solution in the first place.

https://user-images.githubusercontent.com/1482075/134021853-7773f132-d7cb-403a-a950-61c15ddbd5f7.mp4

jasmussen commented:
Not necessarily. In the site editor the block might be flush against the top of the screen, which is why we moved the block toolbar below in #29464. Or you might be inserting the block in a narrow context, such as a column, where the block toolbar, or screen edge, will cover the spinner. 

I do appreciate your point, and I do think it's worth following up with a postponed spinner. I think in many cases, certainly those you demo, there reaction time will be so fast there really is no need to see a spinner at all.

dgwyer commented:
@jasmussen I'll reiterate my point for what it's worth and leave it at that. I would welcome the chance to make this even better for users.

The initial motivation for updating the SSR component was to match as closely the (re)rendering experience of pure JS blocks; which I achieved (with spinner location option). This merged solution does not do that in its current state.

As you pointed out we may not even need the spinner displayed at all in most cases. It would only need to be rendered as a fallback if there is a server delay.

This would be the best all-around solution as it would closely match JS block rendering behavior, fulfilling the original criteria.

BinaryMoon commented:
Hi - I've been following this thread since I like what @dgwyer had done originally and I am in agreement that no dimming is preferable.

As a user experience it looks so much nicer to not fade out, it feels instant even though it takes the same amount of time. I'm disappointed this (appears to be) merged as is since it doesn't feel any better than what we have currently.

jasmussen commented:
I created a new ticket, #35027, to improve the transient state so that it only appears if an operation takes a while. I'll see if I can ask a friend to work on it.
